src/engine/IndexScan.cpp

```diff
@@ -324,15 +341,192 @@ Permutation::IdTableGenerator IndexScan::lazyScanForJoinOfColumnWithScan(
+                                                           metaBlocks.value());
 
   auto result = getLazyScan(blocks);
-  result.details().numBlocksAll_ = metaBlocks1.value().blockMetadata_.size();
+  result.details().numBlocksAll_ = metaBlocks.value().blockMetadata_.size();
   return result;
 }
+
+// _____________________________________________________________________________
+void IndexScan::updateRuntimeInfoForLazyScan(const LazyScanMetadata& metadata) {
+  updateRuntimeInformationWhenOptimizedOut(
+      RuntimeInformation::Status::lazilyMaterialized);
+  auto& rti = runtimeInfo();
+  rti.numRows_ = metadata.numElementsYielded_;
+  rti.totalTime_ = metadata.blockingTime_;
+  rti.addDetail("num-blocks-read", metadata.numBlocksRead_);
+  rti.addDetail("num-blocks-all", metadata.numBlocksAll_);
+  rti.addDetail("num-elements-read", metadata.numElementsRead_);
+
+  // Add more details, but only if the respective value is non-zero.
+  auto updateIfPositive = [&rti](const auto& value, const std::string& key) {
+    if (value > 0) {
+      rti.addDetail(key, value);
+    }
+  };
+  updateIfPositive(metadata.numBlocksSkippedBecauseOfGraph_,
+                   "num-blocks-skipped-graph");
+  updateIfPositive(metadata.numBlocksPostprocessed_,
+                   "num-blocks-postprocessed");
+  updateIfPositive(metadata.numBlocksWithUpdate_, "num-blocks-with-update");
+}
+
+// Store a Generator and its corresponding iterator as well as unconsumed values
+// resulting from the generator.
+struct IndexScan::SharedGeneratorState {
+  Result::Generator generator_;
+  ColumnIndex joinColumn_;
+  Permutation::MetadataAndBlocks metaBlocks_;
+  std::optional<Result::Generator::iterator> iterator_ = std::nullopt;
+  std::optional<Id> lastValue_ = std::nullopt;
+  std::deque<Result::IdTableVocabPair> prefetchedValues_{};
+  std::vector<CompressedBlockMetadata> pendingBlocks_{};
+  bool hasUndef_ = false;
+  bool doneFetching_ = false;
+
+  // Fetch the first value from the generator and skip empty tables.
+  bool fetchFirst() {
+    AD_CORRECTNESS_CHECK(!iterator_.has_value());
+    iterator_ = generator_.begin();
+    while (iterator_ != generator_.end()) {
+      if (iterator_.value()->idTable_.empty()) {
+        ++iterator_.value();
+        continue;
+      }
+      hasUndef_ = iterator_.value()->idTable_.at(0, joinColumn_).isUndefined();
+      return false;
+    }
+    return true;
+  }
+
+  // Consume the next non-empty table from the generator and calculate the next
+  // matching blocks from the index scan.
+  void fetch() {
+    bool exhausted;
+    if (iterator_.has_value()) {
+      AD_CONTRACT_CHECK(iterator_.value() != generator_.end());
+      do {
+        // Skip empty tables
+        ++iterator_.value();
+        exhausted = iterator_.value() == generator_.end();
+      } while (!exhausted && iterator_.value()->idTable_.empty());
+    } else {
+      exhausted = fetchFirst();
```

Comment:
This can be a separate function `advanceInputToFirstNonEmptyTable` that either returns or sets  an `input exhausted` flag.